Add new UserProfileEditPicture component

willdurand opened the pull request:
Fix #4276 – ~⚠️ Depends on #5013 and #5014~

---

## Features

- [x] upload a profile picture
- [x] delete a profile picture

## Gif

![2018-05-17 11 14 25](https://user-images.githubusercontent.com/217628/40168403-a84191dc-59c3-11e8-894c-456e3c870e89.gif)

willdurand commented:
So I guess it is almost ready for a first look @bobsilverberg but there are two PRs before that could be reviewed/merged to reduce the diff here.

In addition, there is also a server issue that makes the feature not entirely finished (see: https://github.com/mozilla/addons-server/issues/8265). I think the only impact would be CSS-related (minor).

I could maybe cut this PR in half to reduce the diff even more!

willdurand commented:
To display the user profile picture (since it does not work because of https://github.com/mozilla/addons-server/issues/8265), I use this patch:

```diff
diff --git a/src/ui/components/UserAvatar/index.js b/src/ui/components/UserAvatar/index.js
index cc28a0b27..f90b15fe8 100644
--- a/src/ui/components/UserAvatar/index.js
+++ b/src/ui/components/UserAvatar/index.js
@@ -20,7 +20,7 @@ const UserAvatar = ({ altText, className, preview, user }: Props) => {

   if (user && !image) {
     image = user.picture_type && user.picture_type.length && user.picture_url
-      && user.picture_url.length ? user.picture_url : null;
+      && user.picture_url.length ? user.picture_url.replace('.png', '_original.png') : null;
   }

   return (

```

willdurand commented:
I have updated the Gif, and fixed the CSS for the delete button. This is good for a review!

bobsilverberg commented on tests/unit/amo/api/test_users.js:
```diff
@@ -134,4 +135,33 @@ describe(__filename, () => {
   });
+
+  describe('deleteUserPicture', () => {
+    const mockResponse = () => createApiResponse({
```
Why is this a function rather than just a variable? Also, I know you've only added the one test, but this `mockResponse` function is used a number of times in this file, so maybe it would be worthwhile to either move it to the top of the file, or have a `createFakeUserAccountResponse` in one of the helper files?

willdurand replied:
mmh right, we could probably move this function to the top of the file.

bobsilverberg commented on tests/unit/amo/api/test_users.js:
```diff
@@ -134,4 +135,33 @@ describe(__filename, () => {
+      jsonData: createUserAccountResponse(),
+    });
+
+    const getParams = (params = {}) => {
```
I think I've brought this up before, but it seems like we're doing a bunch of extra stuff here having a `getParams` function, which can even take an extra set of `params` to override the defaults, but then all we ever do with it is just call `getParams()`. I know it's how we usually do it, but I'm starting to question the value of all that extra code when we never really use it, and could just create the params inline. This is the only test that even has access to `getParams`, and it seems unlikely we'll add any others.

willdurand replied:
We could indeed simplify a bit here.

willdurand commented on src/amo/components/UserProfileEditPicture/index.js:
```diff
@@ -0,0 +1,86 @@
+
+      {(user && user.picture_url) && (
+        <p>
+          <Button
```
Yes, it is intended.

willdurand commented on src/amo/sagas/users.js:
```diff
@@ -90,8 +92,34 @@ export function* fetchUserAccount({
+    errorHandlerId,
+    userId,
+  },
+}) {
```
Yes, definitely: https://github.com/mozilla/addons-frontend/issues/5080

willdurand commented on src/ui/components/UserAvatar/index.js:
```diff
@@ -7,20 +7,33 @@ import type { UserType } from 'amo/reducers/users';
+  let image = preview && preview.length ? preview : null;
+
+  if (user && !image) {
+    image = user.picture_type && user.picture_type.length && user.picture_url
```
I don't know, that was there 🤷‍♂️ I think that's just the way of determining whether there is a profile picture for a given user or not.

bobsilverberg replied:
Maybe it's worth investigating further, or maybe @eviljeff knows? I don't love having mysterious code that we don't fully understand in our code base.

eviljeff replied:
`picture_type` is the mime type of the image [[docs](http://addons-server.readthedocs.io/en/latest/topics/api/accounts.html#account)]
Previously `picture_url` would be a url to a default avatar if there wasn't a user provided picture, so you had to test `picture_type` - which would be `null`/empty if there wasn't a user provided picture - to check if it was set or not.
Since mozilla/addons-server#7679 `picture_url` has been `null` when not set so I don't believe you need that check.

willdurand replied:
Yup, let's forget about `picture_type` then.

bobsilverberg commented on tests/unit/amo/components/TestUserProfilePictureEdit.js:
```diff
@@ -0,0 +1,133 @@
+import * as React from 'react';
```
This file is called `TestUserProfilePictureEdit.js` but the component is called `UserProfileEditPicture`.

willdurand replied:
🤦‍♂️

bobsilverberg commented on tests/unit/amo/components/TestUserProfilePictureEdit.js:
```diff
@@ -0,0 +1,133 @@
+describe(__filename, () => {
+  const render = ({ i18n = fakeI18n(), ...props } = {}) => {
+    return shallowUntilTarget(
+      <UserProfileEditPicture i18n={i18n} name="input-name" {...props} />,
```
This component has some mandatory props that are not being passed. I guess the tests don't care, and I honestly don't recall if we usually worry about that or not, but I wonder if we should have default props for everything that is mandatory?

willdurand replied:
That would make sense, thanks!

bobsilverberg commented on tests/unit/amo/components/TestUserProfilePictureEdit.js:
```diff
@@ -0,0 +1,133 @@
+    expect(root.find('.UserProfileEditPicture-file')).toHaveLength(1);
+    expect(root.find('.UserProfileEditPicture-file-input'))
+      .toHaveProp('name', name);
+    expect(root.find('.UserProfileEditPicture-delete-button')).toHaveLength(0);
```
Further to my comment above, it seems like we're relying on a non-explicit default state for the component in these tests. For example here we expect there to be no delete button, but that's only because of a specific state, which we didn't set up in the test. It seems like it would be better to set up the component state in the test (by passing explicit props).

bobsilverberg commented on tests/unit/amo/components/TestUserProfilePictureEdit.js:
```diff
@@ -0,0 +1,133 @@
+    const root = render();
+
+    expect(root.find(UserAvatar)).toHaveLength(1);
+    expect(root.find(UserAvatar)).toHaveProp('altText', null);
```
Same here, the test doesn't explicitly set itself up for `altText` to be null.

bobsilverberg commented on tests/unit/amo/components/TestUserProfilePictureEdit.js:
```diff
@@ -0,0 +1,133 @@
+    const root = render({ user });
+
+    expect(root.find(UserAvatar))
+      .toHaveProp('altText', `Profile picture for ${user.name}`);
```
We might as well also test that the `UserAvatar` component received the expected `preview` and `user` props as well.

willdurand commented:
@bobsilverberg thanks for the first batch of comments, I have updated the code, fixed a few things and rebased. You can have a fresh look if you wish.

bobsilverberg commented on tests/unit/amo/components/TestUserProfilePictureEdit.js:
```diff
@@ -0,0 +1,133 @@
+
+    const root = render({ user });
+
+    expect(root.find('.UserProfileEditPicture-delete-button')).toHaveLength(1);
```
Is it worth checking the text of the button as well?

bobsilverberg commented on tests/unit/ui/components/TestUserAvatar.js:
```diff
@@ -62,4 +61,12 @@ describe(__filename, () => {
+    expect(root.find('.UserAvatar-image')).toHaveLength(1);
+    expect(root.find('.UserAvatar-image')).toHaveProp('src', preview);
+  });
 });
```
You added an `altText` prop to `UserAvatar` but it doesn't seem to be tested in here.

willdurand commented on tests/unit/amo/components/TestUserProfileEdit.js:
```diff
@@ -697,4 +704,90 @@ describe(__filename, () => {
+      },
+    });
+
+    expect(root.find(Notice)).toHaveLength(1);
```
yes, thanks!

willdurand commented:
Thanks for the review @bobsilverberg, looks like I overlooked a bunch of assertions. This is ready for another look.
